### PR TITLE
Reset password

### LINKED ISSUE #10: User resets their password
As a user, 
I want the ability to create a new password when I’ve forgotten it
In order to resume use of the site.

Acceptance Criteria: 

Ability to lookup account by email
Send email verification with link to new password page
Ability to input new password
New password should be saved to Database

### PR DESCRIPTION
**Pull request by CarterNelms**
resolves #10

### CODE REVIEW

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  def create
+    @user = User.find_by_email(params[:user][:email])
+    # binding.pry
+
+    # This line sends an email to the user with instructions on how to reset their password (a url with a random token)
+    if @user
+      if @user.deliver_reset_password_instructions!
+        redirect_to(root_path, :notice => 'Instructions have been sent to your email.')
+      else
+        redirect_to(root_path, :alert => 'Email was not successfully sent.')
```
Comment (elizamarcum) [inline app/controllers/password_resets_controller.rb]: This message should be more user friendly. e.g. "We could not send your password reset email."

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      end
+    else
+      redirect_to(root_path, :alert => "Couldn't find user.")
```
Comment (elizamarcum) [inline app/controllers/password_resets_controller.rb]: This message should match the message above.  As a security precaution, we don't want to alert hackers that a particular email is/is not registered.

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  def index
+    @user=User.new
+  end
+
+  # request password reset.
+  # you get here when the user entered his email in the reset password form and submitted it.
+  def create
+    @user = User.find_by_email(params[:user][:email])
+    # binding.pry
```
Comment (elizamarcum) [inline app/controllers/password_resets_controller.rb]: This line shouldn't be here.

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
+class PasswordResetsController < ApplicationController
+  skip_before_filter :require_login
+
+  def index
+    @user=User.new
+  end
+
+  # request password reset.
+  # you get here when the user entered his email in the reset password form and submitted it.
```
Comment (elizamarcum) [inline app/controllers/password_resets_controller.rb]: Please remove these comments.  In rails we don't typically comment code unless it is particularly confusing.

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    # This is to not leak information to attackers about which emails exist in the system.
+    
+  end
+
+  # This is the reset password form.
+  def edit
+    @user = User.load_from_reset_password_token(params[:id])
+    @token = params[:id]
+
+    if @user.blank?
```
Comment (elizamarcum) [inline app/controllers/password_resets_controller.rb]: This `@user` loading and `@user.blank?` check is repeated twice.  It should be pulled out into a `before_action :load_user` that is called before both the edit and update actions.  Within the `load_user` method, redirecting will work.  So, you could redirect to root with the message of "Your password reset link is not active.  Please try resetting your password again"

File: app/views/password_resets/edit.html.haml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+%h1 Choose A New Password
+
+/ %form
```
Comment (elizamarcum) [inline app/views/password_resets/edit.html.haml]: This code seems to be commented out.

File: app/views/user_sessions/new.html.haml [hunk 7]
```diff
@@ -2,4 +2,5 @@
 = simple_form_for @user, url: user_session_path, method: :post do |f|
   = f.input :email
   = f.input :password
+  = link_to 'Reset your password', password_resets_path
```
Comment (elizamarcum) [inline app/views/user_sessions/new.html.haml]: This link should go after the form.  I would prefer the wording to be "Forgot password?"

File: config/routes.rb [hunk 8]
```diff
@@ -11,6 +11,7 @@
   resources :gardens, except: [:new]
   resource :user_session, only: [:new, :create, :destroy]
   resource :user, except: [:destroy]
+  resources :password_resets
```
Comment (elizamarcum) [inline config/routes.rb]: I believe that the password resets doesn't have an index, show, or destroy method.  The routes should reflect that.

File: ers [hunk 9]
```diff
@@ -0,0 +1,14 @@
+               List of relations
+ Schema |       Name        |   Type   | Owner
```
Comment (elizamarcum) [inline ers]: Where did this file come from?  I don't believe it belongs here.

File: spec/controllers/password_resets_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe PasswordResetsController, :type => :controller do
```
Comment (elizamarcum) [inline spec/controllers/password_resets_controller_spec.rb]: This file should be deleted, as it isn't used.

File: spec/features/user_resets_password_spec.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
+feature "User resets password" do
+  scenario 'Happy path, reset password and log in' do
+    @user = Fabricate(:user)
+    visit '/'
+    click_link 'Sign In'
+    click_link 'Reset your password'
+    fill_in "Email", with: @user.email
+    click_on "Send Email"
+    @user = User.find_by_email(@user.email)
+    visit "/password_resets/#{@user.reset_password_token}/edit"
```
Comment (elizamarcum) [inline spec/features/user_resets_password_spec.rb]: You should use the rails path helpers to generate the urls in the tests.  e.g. `edit_password_reset_path(@user.reset_password_token)`

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -7,6 +7,10 @@
   it { should validate_presence_of :email }
   it { should validate_presence_of :password }
   it { should validate_uniqueness_of :email }
-
   it { should have_many :gardens }
+
+  let!(:user) { Fabricate(:user) }
+  it "sends an email" do
```
Comment (elizamarcum) [inline spec/models/user_spec.rb]: This should be nested in it's own context, since "sends an email" isn't explanatory of what you're trying to test here.

**Comment by elizamarcum** [comment]
@wesmangum and @CarterNelms: I've noted a number of changes that I'd like you to make.

File: app/controllers/password_resets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,66 @@
+class PasswordResetsController < ApplicationController
+  skip_before_filter :require_login
+
+  def index
+    @user=User.new
```
Comment (wesmangum) [inline app/controllers/password_resets_controller.rb:6]: Accidentally deleted @elizabrock 's comment:

This should be a new method, not index.

**Comment by wesmangum** [comment]
@elizabrock fixed code where you left comments. Please review and add additional comments where you find issues.

**Pull request merged**